# Урок 2. Задания.

### Задание 1
Поработаем с данными о поездках на такси в Нью-Йорке, в которых также имеется информация о погодных условиях и выходных днях.  

pickup_dt – период с точностью до часа

pickup_month – месяц

borough – район Нью-Йорка, из которого был сделан заказ (5 районов + аэропорт)

pickups – число поездок за период (час)

hday – является ли день праздничным/выходным; Y – да, N – нет

spd – скорость ветра в милях в час

vsb – видимость

temp – температура в градусах Фаренгейта

dewp – точка росы по Фаренгейту

slp – давление

pcp_01 – количество осадков за час

pcp_06 – количество осадков за 6 часов

pcp_24 – количество осадков за 24 часа

sd – глубина снега в дюймах

Импортируйте библиотеку pandas, а затем загрузите датасет 2_taxi_nyc, находящийся здесь. Результат сохраните в переменную taxi.

In [2]:
import pandas as pd

In [3]:
taxi = pd.read_csv('https://stepik.org/media/attachments/lesson/360340/2_taxi_nyc.csv')

In [4]:
taxi

pickup_dt pickup_month        borough  pickups hday  spd  \
0      2015-01-01 01:00:00          Jan          Bronx      152    Y  5.0   
1      2015-01-01 01:00:00          Jan       Brooklyn     1519    Y  5.0   
2      2015-01-01 01:00:00          Jan            EWR        0    Y  5.0   
3      2015-01-01 01:00:00          Jan      Manhattan     5258    Y  5.0   
4      2015-01-01 01:00:00          Jan         Queens      405    Y  5.0   
...                    ...          ...            ...      ...  ...  ...   
29096  2015-06-30 23:00:00          Jun            EWR        0    N  7.0   
29097  2015-06-30 23:00:00          Jun      Manhattan     3828    N  7.0   
29098  2015-06-30 23:00:00          Jun         Queens      580    N  7.0   
29099  2015-06-30 23:00:00          Jun  Staten Island        0    N  7.0   
29100  2015-06-30 23:00:00          Jun            NaN        3    N  7.0   

        vsb  temp  dewp     slp  pcp 01  pcp 06  pcp 24   sd  
0      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
1      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
2      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
3      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
4      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
...     ...   ...   ...     ...     ...     ...     ...  ...  
29096  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29097  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29098  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29099  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29100  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  

[29101 rows x 14 columns]

### Задание 2
Проверьте, сколько всего строк и столбцов имеется в датасете.

In [4]:
taxi.shape

(29101, 14)

Cтроки: 29101
Столбцы: 14

### Задание 3
Давайте посмотрим на типы колонок. Все ли из них считались правильно? В качестве ответа выберите тип, преобладающий в датасете.

In [6]:
taxi.dtypes.value_counts()

float64    9
object     4
int64      1
dtype: int64

object - "неопределенный" тип данных

Преобладающий тип - float64

### Задание 4
В названиях столбцов pcp 01, pcp 06, pcp 24 встречается пробел. Это не очень удобно, если в дальнейшем вы планируете обращаться к колонкам через точку, без использования кавычек и скобочек. 

Замените пробел в названиях на знак нижнего подчёркивания. 

Данные сохранены в переменную taxi. Данные должны быть изменены в этом же датафрейме.



In [10]:
# Вариант 1

taxi = taxi.rename(columns = {'pcp 01': 'pcp_01',
                             'pcp 06': 'pcp_06',
                             'pcp 24': 'pcp_24'})

In [11]:
# Вариант 2

def undescore_rename(name):
    return name.replace(' ', '_')

taxi = taxi.rename(columns = undescore_rename)

### Задание 5
Вас попросили узнать, сколько записей (строк) в датафрейме относятся к району Манхэттен (Manhattan). Получить ответ на данный вопрос можно было бы, например, с помощью следующей команды:

taxi.query("borough == 'Manhattan'").shape[0]
А если мы хотим посмотреть, сколько раз встречается каждый из районов? Неужели придётся использовать подобную конструкцию для каждого уровня переменной? Здесь на помощь приходит метод value_counts(), подробнее о котором можно почитать в конспекте! 

А теперь вопрос: сколько раз в данных встречается район Бруклин (Brooklyn)?

In [15]:
taxi.borough.value_counts()

Bronx            4343
Brooklyn         4343
EWR              4343
Manhattan        4343
Queens           4343
Staten Island    4343
Name: borough, dtype: int64

Ответ: 4343

### Задание 6
Следующая задача: выяснить, из какого района было совершено наибольшее количество поездок за весь период. 

Для начала посчитайте общее количество поездок (pickups) без группировки.

In [18]:
taxi.pickups.sum()

14265773

### Задание 7
А теперь сгруппируйте данные по району (borough) и укажите, из какого пункта было совершено больше всего поездок

In [22]:
taxi.groupby('borough', as_index = False)\
    .agg({'pickups': 'sum'})\
    .sort_values('pickups', ascending = False)

borough   pickups
3      Manhattan  10367841
1       Brooklyn   2321035
4         Queens   1343528
0          Bronx    220047
5  Staten Island      6957
2            EWR       105

Ответ: Manhattan

### Задание 8
Ещё несколько полезных методов в pandas! Возможно, в предыдущем степе вы просто посмотрели на данные или отсортировали значения. Ускорить данный процесс в дальнейшем помогут методы idxmin() и idxmax(), которые возвращают индекс минимального или максимального значения.


idxmin – индекс минимального значения

idxmax – индекс максимального значения

Сохраните название района с наименьшим числом поездок в переменную min_pickups, применив подходящий метод.

In [26]:
min_pickups = taxi.groupby('borough')\
                    .agg({'pickups': 'sum'})\
                    .idxmin()

In [27]:
min_pickups

pickups    EWR
dtype: object

### Задание 9
Продолжим изучение данных и посмотрим на число поездок в выходные дни. Сгруппируйте данные по двум признакам: району города и является ли день выходным (колонки borough и hday). Сравните среднее число поездок, и выберите районы, из которых по праздникам в среднем поступает больше заказов, чем в обычные дни. 

In [30]:
taxi.groupby(['borough', 'hday']).\
    agg({'pickups': 'mean'})

pickups
borough       hday             
Bronx         N       50.771073
              Y       48.065868
Brooklyn      N      534.727969
              Y      527.011976
EWR           N        0.023467
              Y        0.041916
Manhattan     N     2401.302921
              Y     2035.928144
Queens        N      308.899904
              Y      320.730539
Staten Island N        1.606082
              Y        1.497006

Ответ: EWR, Queens

### Задание 10
Для каждого района посчитайте число поездок по месяцам. Отсортируйте полученные значения по убыванию и сохраните результирующий датафрейм в pickups_by_mon_bor.

Обратите внимание, что итоговый датасет должен состоять из 3-х колонок: pickup_month, borough, pickups.

In [43]:
pickups_by_mon_bor = taxi.groupby(['borough', 'pickup_month']).\
                            agg({'pickups': 'sum'}).\
                            sort_values('pickups', ascending = False)

In [45]:
pickups_by_mon_bor

pickups
borough       pickup_month         
Manhattan     Jun           1995388
              May           1888800
              Feb           1718571
              Mar           1661261
              Apr           1648278
              Jan           1455543
Brooklyn      Jun            482466
              May            476087
              Apr            378095
              Mar            346726
              Feb            328650
              Jan            309011
Queens        Jun            286311
              May            275893
              Mar            219561
              Apr            216857
              Feb            185695
              Jan            159211
Bronx         May             53037
              Jun             49006
              Apr             34617
              Mar             32232
              Feb             28694
              Jan             22461
Staten Island Jun              1673
              May              1517
              Apr              1068
              Mar               975
              Feb               903
              Jan               821
EWR           Jun                29
              May                27
              Mar                14
              Feb                14
              Jan                11
              Apr                10

### Задание 11
Теперь задача посложнее! Попрактикуемся в написании собственных функций и их применении к датафрейму. Поскольку данные о поездках в Нью-Йорке, температура представлена в градусах Фаренгейта. 

Напишите функцию temp_to_celcius, которая получает на вход колонку с температурой в °F и возвращает значения, переведённые в градусы Цельсия.


Пример работы:

> taxi['temp'][:5]

0    30.0
1    30.0
3    30.0
4    30.0
5    30.0
> taxi['temp_C'] = temp_to_celcius(taxi['temp'])
> taxi['temp_C'][:5]

0   -1.111111
1   -1.111111
3   -1.111111
4   -1.111111
5   -1.111111
 

NB! В окошке ниже надо написать только саму функцию. Так как урок посвящён pandas, то и всё содержимое функции должно быть выполнено средствами этой библиотеки - убедитесь, что на выходе она даёт pandas Series.

In [46]:
def temp_to_celcius(F):
    C = ((F - 32) * 5) / 9
    return C

In [48]:
temp_to_celcius(taxi.temp)

0        -1.111111
1        -1.111111
2        -1.111111
3        -1.111111
4        -1.111111
           ...    
29096    23.888889
29097    23.888889
29098    23.888889
29099    23.888889
29100    23.888889
Name: temp, Length: 29101, dtype: float64

! В данном случае нельзя передать в функцию в качестве аргумента массив 
(список)